src/ja/server/scheduler/algorithm.py

```diff
@@ -0,0 +1,115 @@
+from abc import ABC, abstractmethod
+from typing import List, Optional, Tuple
+from ja.common.job import Job
+from ja.server.database.database import ServerDatabase
+from ja.server.database.types.work_machine import WorkMachine
+
+
+class CostFunction(ABC):
+    """
+    A CostFunction is a collection of functions which determine the order in
+    which jobs are scheduled.
+    """
+
+    @abstractmethod
+    def calculate_cost(self, job: Job) -> int:
+        """!
+        Calculate the cost of a job.
+
+        The cost of a job is an integer. Higher values mean higher priority for
+        the job when considering which jobs to schedule.
+
+        @param job The job whose cost should be calculated.
+        @return The calculated cost of the job.
+        """
+
+    @abstractmethod
+    @property
+    def blocking_threshold(self) -> int:
+        """!
+        @return The minimum cost a job can have so that it can block further
+          jobs with smaller cost from being scheduled.
+        """
+
+    @abstractmethod
+    @property
+    def preempting_threshold(self) -> int:
+        """!
+        @return The minimum cost a job can have so that it can preempt already
+          running jobs. Should be at least @blocking_threshold.
+        """
+
+
+class JobDistributionPolicy(ABC):
+    """
+    JobDistributionPolicy is a collection of functions which determine how
+    jobs to be scheduled are to be distributed among work machines.
```

Comment:
Wouldn't it make more sense to make one JobDistributionPolicy super class that has three sub-classes for regular assigning, blocking, and preemption instead of one class with three methods?

Reply:
I did it this way because otherwise some of the returned values won't be used in all cases, but maybe it doesn't matter. I'll change it as you say.